### DISCUSSION TITLE
Lesson 10: LendingPool.withdraw() function

### DISCUSSION THREAD

**Discussion by ChakibHi** [smartcontractkit/full-blockchain-solidity-course-py#30] [Q&A] [answered]
Hello everyone,
at the end of Lesson 10, we created a python script that will do the following:
- Deposit to Aave (using `LendingPool.deposit()`)
- Borrow Dai (using `LendingPool.borrow()`)
- Repay Dai (using `LendingPool.repay()`)

However the first deposit is still locked inside Aave. To withdraw everything and complete the process, LendingPool documentation have a **`withdraw()`** function to withdraw WETH and burn the **aToken**.

![image](https://user-images.githubusercontent.com/40836140/133660519-956e8ef9-ee1b-4f33-8a9e-cd09acbaeb05.png)

The blocking point is, how to withdraw the entire balance (with interest) using python?

FYI, I successfully did it inside a Solidity contract by passing the **"`type(uint).max`"** as an argument, but i couldn't find a way to do it frontend only (python).

**Comment by AlbertoSinigaglia** [comment]
from [this contract](https://etherscan.io/contractdiffchecker?a1=0xd4aa6370deb717ef707f9f62677b90b5df4349d0) I can see a `withdrawAll` method... you probably can do the same with your, and since the amount is retrieved by looking at the balance of an address, you might be able to do the same with Python

In my case, I had to add:
```solidity
function withdraw(uint wad) public;
```
to the interface that @PatrickAlphaC gave us, and then used this code:
```python
contract = interface.IWETHGateway(config["networks"][network.show_active()]["weth_token"])
tx = contract.withdraw(contract.balanceOf(account), {
  "from": account,
})
```

**Reply by PatrickAlphaC** [reply]
@ChakibHi I haven't tested this, but is this accurate?

**Reply by AlbertoSinigaglia** [reply]
@PatrickAlphaC  i was able to reproduce that code using python, however since involves  call and transaction, there might be some race condition if while we do our calculation someone else withdraw of funds the account

**Reply by AlbertoSinigaglia** [reply]
@ChakibHi updated the answer

**Reply by ChakibHi** [reply]
Thanks @AlbertoSinigaglia for the suggested answer! 

However I just found an answer to my question, which was how to send the "`type(uint256).max`" through Web3.py so we can withdraw the whole deposit whatever it was.

PS: The **withdrawETH()** on `IWETHGateway.sol` need three arguments + the interface have its own address "WETHGateway_address" + balanceOf is used normally on the token.

**Reply by AlbertoSinigaglia** [reply]
@ChakibHi wee you able to get all the eth back from Aave? The best i was able to do is to reset the borrowed amount, but I can't find a way to get back the 0.1 eth deposited at the beginning

**Reply by ChakibHi** [reply]
@AlbertoSinigaglia Could you try my answer below and let me know if it works on your end.

**Comment by ChakibHi** [comment]
The `type(uint).max` annotation is basically for Solidity smart contracts, to send the max uint256 using Web3.py or Web3.js we can use one of the following:

* max amount = `2**256 -1`
* max amount = `Web3.toHex(2 ** 256 - 1)  `

I used the WETHGateway, because it easier to withdraw the unwrapped Ether directly (not WETH):

```python
def aave_withdrawEth():
    aave_WETHGateway = interface.IWETHGateway(
        config["networks"][network.show_active()]["WETHGateway_address"]
    )
    tx = aave_WETHGateway.withdrawETH(
        lending_pool_address,
        Web3.toHex(2 ** 256 - 1),  # We want to withdraw the max amount on Aave
        account,
        {"from": account},
    )
    tx.wait(1)
    print("You withdrawn Eth from Aave successfully!")
```



you may need to approve the aWETH ERC20 token.
https://docs.aave.com/developers/the-core-protocol/weth-gateway#withdraweth

**Reply by AlbertoSinigaglia** [reply]
can you also provide the code for: 
> you may need to approve the aWETH ERC20 token.

?

**Reply by ChakibHi** [reply]
Hi @AlbertoSinigaglia , you can use the `approve_erc20()` function that @PatrickAlphaC defined; 

```python
approve_erc20(amount, spender, erc20_address, account)
```
`amount` = max
`erc20_address` = aWETH address

**Reply by AlbertoSinigaglia** [reply]
@ChakibHi  but why should we approve the transfer? I mean, when we deposit eth to aave using `deposithETH` we are moving ETH... and when we withdraw eth, we are working with eth, why should I approve any payment for WETH? at most, It's aave that should approve the withdraw to WETH

**Reply by ChakibHi** [reply]
@AlbertoSinigaglia when you deposit ETH, they send you the equivalent amount in **aWETH** (_Aave interest bearing WETH_) directly to your wallet. According to their documentation, when you withdraw they need the approval to burn the **aWETH** that was sent to you.

![image](https://user-images.githubusercontent.com/40836140/133927420-71a3b674-437e-4cce-87c2-6b4036a8c3c4.png)

**Reply by AlbertoSinigaglia** [reply]
this is not working for me, it keeps saying `ERC20: transfer amount exceeds allowance`

**Reply by AlbertoSinigaglia** [reply]
this is the code I'm ending up using that makes total sense to me, but it's still not working:
```python
def withdrawEthFromAave():
    account = getAccount()
    aweth_token = getAaveAWethGateway()
    lending_pool = getLendingPool()
    amount = aweth_token.balanceOf(account.address)
    print(approveERC20(
        token_contract_address=aweth_token.address,
        spender=lending_pool.address,
        value=amount*1000,
        owner=account
    ).return_value)
    tx = getAaveWethGateway().withdrawETH(
        lending_pool.address,  # lendingPool
        amount,  # amount
        account,  # to
        {"from": account},
    )
    tx.wait(1)
    print("You withdrawn Eth from Aave successfully!")
```

**Reply by PatrickAlphaC** [reply]
Love the thread here, let me just catch up. @AlbertoSinigaglia I think you solved this issue in another thread right? Or maybe I'm getting mixed up. 

Is there a TL;DR of where we are up to this point?

**Comment by PatrickAlphaC** [comment, marked as answer]
Ok now I'm up to speed with these. Amazing questions all.

A few points of confusion.

## There are 2 kinds of `withdraw` functions in this process
 - One is to withdraw your ETH out of the WETH contract. (converting your WETH ERC20 version of ETH _back_ into it's regular non-tokenized version of ETH.
 - Withdrawing your deposited collateral from the Aave contract.

The original question here is asking _"How do I withdraw all of my original deposited collateral?"_

Which we can do by:
1. First repaying all our debts
2. Then withdrawing everything (using that nice -1 syntax)

_Note: This brings up a good point, in a bigger project it would be nice to refactor the code so that we had fewer duplicated functions, but for simplicity, I'm just going to write the whole function. 

### Repay our debts

Now... since we are in fake testnet world, it's sort of hard to get more DAI, it's a bit easier in the forked blockchain world, but basically, you need to buy more DAI to pay it back. I've got some scripts on how to do that here: https://github.com/PatrickAlphaC/defi_py_mix/blob/main/scripts/swap/swap.py

### Withdrawing
Once you do that, you should then be able to withdraw it pretty easily with something like:

```python
def withdraw_all_collateral(lending_pool, account):
    print("Withdrawing collateral...")
    tx = lending_pool.withdraw(
        config["networks"][network.show_active()]["weth_token"],
        (2 ** 256) - 1,
        account,
        {"from": account},
    )
    tx.wait(1)
    print("Withdrew!")
    return tx
``` 

`uint(-1)` in solidity is the same as `(2 ** 256) - 1` in python. 

## WETH withdraw

In WETH, you can just call the `withdraw` function to swap the ERC20 version of ETH back to regular ETH. I have added the `withdraw` function to the interface in the GitHub repo, so you can test building out a function yourself!

**Reply by PatrickAlphaC** [reply]
@AlbertoSinigaglia @ChakibHi let me know if this clears everything up.

**Reply by ChakibHi** [reply]
Yes cristal clear! Thank you @PatrickAlphaC

**Reply by AlbertoSinigaglia** [reply]
yes, thank you

**Reply by Niferu** [reply]
Hello guys,

I have tried above code and indeed it is working for local environment, but when I'm running it for kovan it is throwing me Gas Estimation error even if I have kovanETH, aWETH, WETH and LINK on my account. I was trying to solve it but failed. I have also tried using IWeth withdraw function, but it gives me same error nad it's not working on local. Can i get any tips how to fix those functions? Thank you in advance for help.

Patrick's function:
```python
max_amount = (2 ** 256) - 1

def withdraw_all_collateral(lending_pool, account):
    print("Withdrawing collateral...")
    tx = lending_pool.withdraw(config["networks"][network.show_active()]["weth_token"], max_amount, account, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```
IWeth interface:
```py
def withdraw_weth():
    account = get_account()
    weth = interface.IWeth(config["networks"][network.show_active()]["a_weth_token"])
    print("Withdrawing collateral...")
    tx = weth.withdraw(2 ** 256 - 1, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```

**Reply by cromewar** [reply]
Hello @Neftyr  be sure you have the correct contract addresses on `bronwie-config.yaml` because AAVE has changed a lot since the video.

**Reply by Niferu** [reply]
Hello @cromewar 
I have tested few addresses in `brownie-config.yaml` and none of them worked throwing same error. I think correct address for 'a_weth_token'(aWETH) is '0x030bA81f1c18d280636F32af80b9AAd02Cf0854e' for mainnet and '0x87b1f4cf9BD63f7BBD3eE1aD04E8F52540349347' for kovan. Let me know if I'm wrong. Thank you in advance mate!

**Reply by cromewar** [reply]
Well haven't touched AAVE in while as they have changed a lot, I'll do some research on this, on the meantime could you please clone the repo of that lesson and run it, does it work?

**Reply by Niferu** [reply]
Hello @cromewar 
There is no Kovan network available for any version of AAVE since today or even yesterday, so i think it became deprecated. I will retry with Goerli now. There are no such functions in Patrick's course, so repo clone won't work here. It is something more that I wanted to do as part of my learning.

**Reply by Niferu** [reply]
Hi @cromewar and everyone i have made some researches and:

Below function based on `lending_pool` works perfect and withdrawing everything from AAVE account for both local testnet and goerli testnet, so previous issue with it was indeed address in `brownie-config.yaml` as AAVE use it's own addresses for WETH and aWETH, you can check that by clicking asset you want address of then as it is shown on screen below:

**AAVE token address checking:**
![image](https://user-images.githubusercontent.com/110537615/188132578-f1b443f5-690f-42a3-8b9c-b797e01c5eeb.png)

**Below function based on `lending_pool`:**
```py
max_amount = 2 ** 256 - 1

def withdraw_all_collateral(lending_pool, account):
    print("Withdrawing collateral...")
    tx = lending_pool.withdraw(config["networks"][network.show_active()]["weth_token"], max_amount, account, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```
Below function based on `IWeth` interface has something incorrect with value `2 ** 256 - 1` as it is `uint` instead of `uint256` i guess it should be `2 ** 32 - 1` and it works but not withdrawing any amount from AAVE account. If we put value of for example 0.1 it will withdraw 0.1, but I want to make it `max_amount` as it is in first function. I have also tried changing `IWeth` interface to: `function withdraw(uint256 wad) external;` but it doesn't work at all like this. Any idea how to fix this amount?

**Below function based on `IWeth` interface:**
```py
def withdraw_weth():
    account = get_account()
    weth = interface.IWeth(config["networks"][network.show_active()]["weth_token"])
    print("Withdrawing collateral...")
    tx = weth.withdraw((2 ** 32) - 1, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```

**Comment by Niferu** [comment]
Hello,

I have just discovered that there is discrepancy between those 2 functions below.

**Function based on `lending_pool` is witdrawing from `Supplied ETH` to `WETH`**
```py
max_amount = 2 ** 256 - 1

def withdraw_all_collateral(lending_pool, account):
    print("Withdrawing collateral...")
    tx = lending_pool.withdraw(config["networks"][network.show_active()]["weth_token"], max_amount, account, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```
**Function based on `IWeth` interface is withdrawing from `WETH` to `ETH` (Not Supplied ETH)**
```py
def withdraw_weth():
    account = get_account()
    weth = interface.IWeth(config["networks"][network.show_active()]["weth_token"])
    print("Withdrawing collateral...")
    tx = weth.withdraw((2 ** 32) - 1, {"from": account})
    tx.wait(1)
    print("Withdrew!")
    return tx
```
I'm not really sure why is that...

I also still didn't figure out why `max_amount` doesn't work for function based on `IWeth` interface. If anyone have idea how to fix it I would be really grateful.

**Reply by cromewar** [reply]
The AAVE protocol changed a lot since the video, so most of things would not work in the same way.
I'll try to create a implementation of the latest version and share it here.
